flux-ping: output header before output of main ping output

chu11 opened the pull request:
Per discussion in #4106, output a header before the main ping output.  This is useful because there are so many input possibilities, as well defaults, that the header can make it clear what is being pinged.  In addition, when a hostname is input on the command line, the rank can be specified too.

As discussed, the output of the ping output is now also the same regardless of the input mechanism.  Before you might get `0!broker`, `broker`, or `upstream!broker` depending on the input on the command line.  Now you always get `0!broker` if the ping comes from rank 0.

Biggest complaint might be my static inline `ASPRINTF()` function.  I was just annoyed enough about passing buffers around that I wanted to use `asprintf` instead, and it was just annoying enough to have to do `log_err_exit("out of memory")` enough times I made the function.

chu11 commented:
re-pushed, the rebase went super poorly

garlick commented:
we do have `xasprintf()` in libutil/xzmalloc.h...

chu11 commented:
@garlick oh! nice

grondo reviewed (commented):
Seems like a good improvement. One main comment/question inline.

Also, I wonder if it would be worth making the new member of the ping response optional? The ping RPC has remained unchanged for quite awhile, might be nice to continue to allow the newer ping to work with an older version of flux-core. Now that I type though, it seems really unimportant, but thought I'd mention it anyway.

chu11 commented on src/cmd/flux-ping.c:
```diff
@@ -86,20 +86,27 @@ char *cred_str (uint32_t userid, uint32_t rolemask, char *buf, int bufsz)
     }
 }
 
+char *rank_bang_str (uint32_t rank, char *buf, int bufsz)
+{
+    if (snprintf (buf, bufsz, "%" PRIu32 "!", rank) >= bufsz)
+        return "";
+    return buf;
+}
```
Yeah it isn't necessary.  I made it sort of a "side effect" to the fact that there used to be function call to handle this.  But we could just directly print out the result instead.

grondo replied:
I meant it seems to do the same thing as the previous function named `rank_bang_str()` (now mysteriously renamed `nodeid_bang_str()`) The `default` case in `nodeid_bang_str()` and this function appear to be duplicated code.

Why not just call the original `rank_bang_str()` (which will call the `default` clause for actual ranks) or `rank_bang_str()` could be moved up, and then replace the code in `default` with a call to that function?

chu11 replied:
hmm, I think I renamed just b/c of ...

```
-                      rank_bang_str (ctx->nodeid, rbuf, sizeof (rbuf)),
+                      nodeid_bang_str (ctx->nodeid, rbuf, sizeof (rbuf)),
```

its passed a "nodeid" vs a "rank", but that's just a weird nomenclature thing.  Lemme try with your suggestion.

grondo replied:
Ah, I see, that does make sense

chu11 replied:
oh wait, i'm dumb.  Once we get the response from the ping callback, we'll know rank != ANY/UPSTREAM.  so we can just call the original `rank_bang_str()`.

chu11 commented:
re-pushed, fixing up per comments above.  split out the `xasprintf()` changes in a new commit.  Removed the new `rank_bang_str()` function (which means the old one stayed).  It ended up the addition was completely unnecessary.

grondo commented:
> as I think the odds of it being needed is pretty low. In addition, it would probably be more hassle than its worth to add some testing for such coverage.

That was the first thing I tried and it failed for me :wink:. Keep in mind that it is common practice to launch a newer version of Flux, i.e. for testing, under an older version of Flux (i.e. a system instance), and then if you try to `flux --parent ping`, the command will fail with `EPROTO`. However, I agree this use case is probably pretty low and the problem will mostly fade over one release cycle.

However, breaking protocols like this could become more and more of an issue as time goes on.

grondo reviewed (approved):
LGTM, Thanks!
